provider/resources.go

```diff
@@ -317,65 +321,105 @@ func nameField(info *tfbridge.SchemaInfo) map[string]*tfbridge.SchemaInfo {
+	if err != nil {
+		return nil, fmt.Errorf("failed to create compute service: %w", err)
 	}
-	project := tfbridge.ConfigStringValue(vars, "project", []string{
-		"GOOGLE_PROJECT",
-		"GOOGLE_CLOUD_PROJECT",
-		"GCLOUD_PROJECT",
-		"CLOUDSDK_CORE_PROJECT",
-	})
-	if project == "" {
-		msg := "unable to detect a global setting for GCP Project;\n" +
-			"Pulumi will rely on per-resource settings for this operation.\n" +
-			"Set the GCP Project by using:\n" +
-			"\t`pulumi config set gcp:project <project>`"
-		host.Log(ctx, diag.Warning, "", msg) // the URN will default to the root stack name which is exactly what we want
-		//return fmt.Errorf("unable to find required configuration setting: GCP Project\n" +
-		//"Set the GCP Project by using:\n" +
-		//"\t`pulumi config set gcp:project <project>`")
-		//
-		return nil
+	regionsService := compute.NewRegionsService(computeService)
+	regionList, err := regionsService.List(project).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list regions: %w", err)
 	}
+	var regions []string
+	for _, region := range regionList.Items {
+		regions = append(regions, region.Name)
+	}
+	return regions, nil
+}
 
-	config := tpg_transport.Config{
-		AccessToken: tfbridge.ConfigStringValue(vars, "accessToken", []string{"GOOGLE_OAUTH_ACCESS_TOKEN"}),
-		Credentials: tfbridge.ConfigStringValue(vars, "credentials", []string{
-			"GOOGLE_CREDENTIALS",
-			"GOOGLE_CLOUD_KEYFILE_JSON",
-			"GCLOUD_KEYFILE_JSON",
-		}),
-		ImpersonateServiceAccount: tfbridge.ConfigStringValue(vars, "impersonateServiceAccount", []string{"GOOGLE_IMPERSONATE_SERVICE_ACCOUNT"}),
-		Project: tfbridge.ConfigStringValue(vars, "project", []string{
+//go:embed errors/no_credentials.txt
+var noCredentialsErr string
+
+//go:embed errors/wrong_region.txt
+var wrongRegionErr string
+
+//go:embed errors/no_project.txt
+var noProjectErr string
+
+func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpClientOpts []option.ClientOption) func(
+	ctx context.Context, host *provider.HostClient, vars resource.PropertyMap, c shim.ResourceConfig,
+) error {
+	return func(ctx context.Context, host *provider.HostClient, vars resource.PropertyMap, c shim.ResourceConfig) error {
+		if !credentialsValidationRun.CompareAndSwap(false, true) {
+			return nil
+		}
+		project := tfbridge.ConfigStringValue(vars, "project", []string{
 			"GOOGLE_PROJECT",
 			"GOOGLE_CLOUD_PROJECT",
 			"GCLOUD_PROJECT",
 			"CLOUDSDK_CORE_PROJECT",
-		}),
-		Region: tfbridge.ConfigStringValue(vars, "region", []string{
-			"GOOGLE_REGION",
-			"GCLOUD_REGION",
-			"CLOUDSDK_COMPUTE_REGION",
-		}),
-		Zone: tfbridge.ConfigStringValue(vars, "zone", []string{
-			"GOOGLE_ZONE",
-			"GCLOUD_ZONE",
-			"CLOUDSDK_COMPUTE_ZONE",
-		}),
-	}
+		})
+		// host is nil in tests.
+		if project == "" && host != nil {
+			host.Log(ctx, diag.Warning, "", noProjectErr) // the URN will default to the root stack name which is exactly what we want
+		}
```

Comment:
This early return means that we don't check credentials if no global project is enabled. I am not sure why it is there but have returned it.

Reply:
That is the goal, yes. It was there previously, and I'm not confident enough to remove it... 

That said, it makes sense. You might have a provider config that is only valid when set to the project you are using. Previously, you could specify the project at the resource level and it would work. Changing to require the project at the provider level is a breaking change. 


The way I think about it, a $`config`$ is valid if it is in this set:

$`valid = \{ config = (AccessToken, Credentials, ImpersonateServiceAccount, Project, Region, Zone) | LoadAndValidate(config) == Valid\}`$

By allowing validation to proceed when project tis not set, we restrict the set of configs we can process to 

$`\{ config = (AccessToken, Credentials, ImpersonateServiceAccount, \"\", Region, Zone) | LoadAndValidate(config) == Valid\} \subset valid`$

We don't want to impose this on our users.